features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,993 @@
+    // Shared lock is in order to prevent GC from operating (which uses exclusive lock).
+    _lock.shared_lock();
+
+    save_active_area = _active_area;
+    record_size = align_up(sizeof(record_header_t) + buf_len_bytes, FLASH_MINIMAL_PROG_UNIT);
+
+    // Parallel operation of writers is allowed due to this atomic operation. This operation
+    // produces an offset on which each writer can work separately, without being interrupted
+    // by the other writer. The only mutual resource here is _free_space_offset - which
+    // gets the correct value because of this atomic increment.
+    new_free_space = safe_increment(_free_space_offset, record_size);
+    record_offset = new_free_space - record_size;
+
+retry:
+    // If we cross the area limit, we need to invoke GC. However, we should consider all the cases
+    // where writers work in parallel, and we only want the FIRST writer to invoke GC.
+    if (new_free_space >= _size) {
+        // In the case we have crossed the limit, but the initial offset was still before the limit, this
+        // means we are the first writer (common case). Exclusively lock _lock, and invoke GC.
+        if (record_offset < _size) {
+            _lock.promote();
+            ret = garbage_collection(key, flags, buf_len_bytes, buf);
+            _lock.exclusive_unlock();
+            return ret;
+        }
+        else {
+            // In the case we have crossed the limit, and the initial offset was also after the limit,
+            // this means we are not the first writer (uncommon case). Just wait for GC to complete.
+            // then retry the operation
+            _lock.shared_unlock();
+            for (;;) {
+                rtos::Thread::wait(MEDITATE_TIME_MS);
+                // Retry operation
+                _lock.shared_lock();
+                new_free_space = safe_increment(_free_space_offset, record_size);
+                goto retry;
+            }
+        }
+    }
+
+    // Now write the record
+    ret = write_record(_active_area, record_offset, key, flags, buf_len_bytes, buf, next_offset);
+    if (ret != NVSTORE_SUCCESS) {
+        _lock.shared_unlock();
+        return ret;
+    }
+
+    // Update _offset_by_key. High bit indicates area.
+    if (flags & DELETE_ITEM_FLAG)
+        _offset_by_key[key] = 0;
+    else
+        _offset_by_key[key] = record_offset | (_active_area << OFFS_BY_KEY_AREA_BIT_POS) |
+                                (((flags & SET_ONCE_FLAG) != 0) << OFFS_BY_KEY_SET_ONCE_BIT_POS);
+
+    _lock.shared_unlock();
+
+    return NVSTORE_SUCCESS;
+}
+
+int NVStore::set(uint16_t key, uint16_t buf_len_bytes, const uint32_t *buf)
+{
+    return do_set(key, buf_len_bytes, buf, 0);
+}
+
+int NVStore::set_once(uint16_t key, uint16_t buf_len_bytes, const uint32_t *buf)
+{
+    return do_set(key, buf_len_bytes, buf, SET_ONCE_FLAG);
+}
+
+int NVStore::remove(uint16_t key)
+{
+    return do_set(key, 0, NULL, DELETE_ITEM_FLAG);
+}
+
+int NVStore::init()
```

Comment:
Can you please review and see if we can optimize the stack usage of this function?

Reply:
Yes, the local variables aren't optimally arranged (alignment-wise). After arrangement they'll use 55 bytes of stack (all arrays there have two indexes). Is this problematic? Note that this is the init function, which should be called once at system life time.